finality-aleph/src/data_io/data_store.rs

```diff
@@ -309,6 +309,27 @@ where
                             debug!(target: "aleph-data-store", "Requesting a stale block {:?} after it has been missing for {:?} secs.", block, time_waiting.as_secs());
                             self.block_requester
                                 .request_stale_block(block.hash, block.num);
+                        } else {
+                            // The whole branch has been imported, what's holding us must be that the parent of the base is not finalized
+                            let bottom_block = proposal.bottom_block();
+                            if let Ok(parent_hash) =
+                                self.chain_info_provider.get_parent_hash(&bottom_block)
+                            {
+                                let parent_num = bottom_block.num - NumberFor::<B>::one();
+                                if self
+                                    .chain_info_provider
+                                    .get_finalized_at(parent_num)
+                                    .is_err()
+                                {
+                                    debug!(target: "aleph-data-store", "Requesting a justification for block {:?} {:?} after it has been missing for {:?} secs.", parent_num, parent_hash, time_waiting.as_secs());
+                                    self.block_requester
+                                        .request_justification(&parent_hash, parent_num);
+                                } else {
+                                    warn!(target: "aleph-data-store", "The proposal {:?} is pending even though blocks were imported and parent was finalized", proposal);
+                                }
+                            } else {
+                                warn!(target: "aleph-data-store", "Expected the block below the proposal {:?} to be imported", proposal);
+                            }
```

Comment:
this is a comment to the whole for loop section but github doesn't allow commenting on the unchanged lines :/

nit: to reduce nesting:
1.
```rust
            if self.bump_proposal(&proposal) {
                continue;
            }
```

2.
``` rust
            let time_waiting = match now.duration_since(first_occurrence) {
                Ok(tw) if tw >= self.config.request_block_after => tw,
                _ => continue,
            };
```

3.
```rust
            if !self.chain_info_provider.is_block_imported(&block) {
                debug!(target: "aleph-data-store", "Requesting a stale block {:?} after it has been missing for {:?} secs.", block, time_waiting.as_secs());
                self.block_requester
                    .request_stale_block(block.hash, block.num);
                continue;
            }
```

4.
```rust
            let parent_hash = match self.chain_info_provider.get_parent_hash(&bottom_block)
            {
                Ok(ph) => ph,
                _ =>  {
                    warn!(target: "aleph-data-store", "Expected the block below the proposal {:?} to be imported", proposal);
                    continue;
                }
            };
```

1 -> line 303
2 -> lin3 305
3 -> line 308
4 -> line 315

Reply:
Addressed -- everything super-flat now. Like Earth.